Add Check Support for Java 21 Record Pattern Syntax : UnnecessaryParenthesesCheck

child of #14890 
I have read check documentation: https://checkstyle.org/checks/coding/unnecessaryparentheses.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words
```
PS D:\CS\test> cat src/Test2.java  
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test2 {
    void test(Object obj) {
        if (!(obj instanceof Rectangle(_, _))) { }  // ok, expected
        if ((!(obj instanceof Rectangle(_, _)))) { }  // violation expected
        if (obj instanceof (Rectangle(_,_))) { }  // UnnecessaryParentheses around the pattern will not compile
    }

}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="UnnecessaryParentheses">

        </module>

    </module>
</module>
PS D:\CS\test> javac src/Test2.java                                              
src\Test2.java:7: error: illegal start of type
        if (obj instanceof (Rectangle(_,_))) { }  // UnnecessaryParentheses around the pattern will not compile
                           ^
1 error
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test2.java
Starting audit...
[ERROR] D:\CS\test\src\Test2.java:6:13: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```

--- 
**Describe what you expect in detail.**

This can be closed without any updates

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
@@ -185,7 +185,11 @@ public class UnnecessaryParenthesesCheckTest extends AbstractModuleTestSupport {
 
     @Test
     public void testUnnecessaryParenthesesPatterns() throws Exception {
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        final String[] expected = {
+            "24:22: " + getCheckMessage(MSG_ASSIGN),
+            "27:21: " + getCheckMessage(MSG_ASSIGN),
+            "31:13: " + getCheckMessage(MSG_EXPR),
+        };
         verifyWithInlineConfigParser(
                 getNonCompilablePath(
                 "InputUnnecessaryParenthesesCheckPatterns.java"),

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckPatterns.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesCheckPatterns.java
@@ -11,7 +11,7 @@ tokens = (default)EXPR, IDENT, NUM_DOUBLE, NUM_FLOAT, NUM_INT, NUM_LONG, \
 
 */
 
-//non-compiled with javac: Compilable with Java17
+//non-compiled with javac: Compilable with Java21
 package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
 
 public class InputUnnecessaryParenthesesCheckPatterns {
@@ -20,5 +20,19 @@ public class InputUnnecessaryParenthesesCheckPatterns {
 
         boolean result = (o instanceof String a) ?
                 (o instanceof String x) : (!(o instanceof String y));
+
+        boolean f =  (o instanceof String x);
+        // violation above, 'Unnecessary parentheses around assignment right-hand side.'
+
+        boolean b = (!(o instanceof Rectangle(_, _)));
+        // violation above, 'Unnecessary parentheses around assignment right-hand side.'
+        boolean c = !(o instanceof Rectangle(_, _));
+
+        if ((!(o instanceof Rectangle(_, _)))) {}
+        // violation above, 'Unnecessary parentheses around expression.'
+        if (!(o instanceof Rectangle( _, _))) {}
+
     }
+
+    record Rectangle(int x, int y) { }
 }

